File: partiql-ast/src/main/java/org/partiql/ast/v1/Ast.java
package org.partiql.ast.v1;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.partiql.ast.v1.expr.Expr;
import org.partiql.ast.v1.expr.ExprAnd;
import org.partiql.ast.v1.expr.ExprArray;
import org.partiql.ast.v1.expr.ExprBag;
import org.partiql.ast.v1.expr.ExprBetween;
import org.partiql.ast.v1.expr.ExprCall;
import org.partiql.ast.v1.expr.ExprCase;
import org.partiql.ast.v1.expr.ExprCast;
import org.partiql.ast.v1.expr.ExprCoalesce;
import org.partiql.ast.v1.expr.ExprExtract;
import org.partiql.ast.v1.expr.ExprInCollection;
import org.partiql.ast.v1.expr.ExprIsType;
import org.partiql.ast.v1.expr.ExprLike;
import org.partiql.ast.v1.expr.ExprLit;
import org.partiql.ast.v1.expr.ExprMatch;
import org.partiql.ast.v1.expr.ExprNot;
import org.partiql.ast.v1.expr.ExprNullIf;
import org.partiql.ast.v1.expr.ExprOperator;
import org.partiql.ast.v1.expr.ExprOr;
import org.partiql.ast.v1.expr.ExprOverlay;
import org.partiql.ast.v1.expr.ExprParameter;
import org.partiql.ast.v1.expr.ExprPath;
import org.partiql.ast.v1.expr.ExprPosition;
import org.partiql.ast.v1.expr.ExprQuerySet;
import org.partiql.ast.v1.expr.ExprSessionAttribute;
import org.partiql.ast.v1.expr.ExprStruct;
import org.partiql.ast.v1.expr.ExprSubstring;
import org.partiql.ast.v1.expr.ExprTrim;
import org.partiql.ast.v1.expr.ExprValues;
import org.partiql.ast.v1.expr.ExprVarRef;
import org.partiql.ast.v1.expr.ExprVariant;
import org.partiql.ast.v1.expr.ExprWindow;
import org.partiql.ast.v1.expr.PathStep;
import org.partiql.ast.v1.expr.Scope;
import org.partiql.ast.v1.expr.SessionAttribute;
import org.partiql.ast.v1.expr.TrimSpec;
import org.partiql.ast.v1.expr.WindowFunction;
import org.partiql.ast.v1.graph.GraphDirection;
import org.partiql.ast.v1.graph.GraphLabel;
import org.partiql.ast.v1.graph.GraphMatch;
import org.partiql.ast.v1.graph.GraphPart;
import org.partiql.ast.v1.graph.GraphPattern;
import org.partiql.ast.v1.graph.GraphQuantifier;
import org.partiql.ast.v1.graph.GraphRestrictor;
import org.partiql.ast.v1.graph.GraphSelector;
import org.partiql.value.PartiQLValue;

import java.util.List;
import java.util.Map;

// TODO docs for all factory methods
public class Ast {
    // Expr
    @NotNull
    public static ExprAnd exprAnd(Expr lhs, Expr rhs) {
        return new ExprAnd(lhs, rhs);
    }

    @NotNull
    public static ExprArray exprArray(@NotNull List<Expr> values) {
        return new ExprArray(values);
    }

    @NotNull
    public static ExprBag exprBag(@NotNull List<Expr> values) {
        return new ExprBag(values);
    }

    @NotNull
    public static ExprBetween exprBetween(@NotNull Expr value, @NotNull Expr from, @NotNull Expr to, boolean not) {
        return new ExprBetween(value, from, to, not);
    }

    @NotNull
    public static ExprCall exprCall(@NotNull IdentifierChain function, @NotNull List<Expr> args, @Nullable SetQuantifier setq) {
        return new ExprCall(function, args, setq);
    }

    @NotNull
    public static ExprCase exprCase(@Nullable Expr expr, @NotNull List<ExprCase.Branch> branches, @Nullable Expr defaultExpr) {
        return new ExprCase(expr, branches, defaultExpr);
    }

    @NotNull
    public static ExprCase.Branch exprCaseBranch(@NotNull Expr condition, @NotNull Expr expr) {
        return new ExprCase.Branch(condition, expr);
    }

    @NotNull
    public static ExprCast exprCast(@NotNull Expr value, @NotNull DataType asType) {
        return new ExprCast(value, asType);
    }

    @NotNull
    public static ExprCoalesce exprCoalesce(@NotNull List<Expr> args) {
        return new ExprCoalesce(args);
    }

    @NotNull
    public static ExprExtract exprExtract(@NotNull DatetimeField field, @NotNull Expr source) {
        return new ExprExtract(field, source);
    }

    @NotNull
    public static ExprInCollection exprInCollection(@NotNull Expr lhs, @NotNull Expr rhs, boolean not) {
        return new ExprInCollection(lhs, rhs, not);
    }

    @NotNull
    public static ExprIsType exprIsType(@NotNull Expr value, @NotNull DataType type, boolean not) {
        return new ExprIsType(value, type, not);
    }

    @NotNull
    public static ExprLike exprLike(@NotNull Expr value, @NotNull Expr Pattern, @Nullable Expr escape, boolean not) {
        return new ExprLike(value, Pattern, escape, not);
    }

    // This representation will be changed in https://github.com/partiql/partiql-lang-kotlin/issues/1589
    @NotNull
    public static ExprLit exprLit(@NotNull PartiQLValue value) {
        return new ExprLit(value);
    }

    @NotNull
    public static ExprMatch exprMatch(@NotNull Expr expr, @NotNull GraphMatch pattern) {
        return new ExprMatch(expr, pattern);
    }

    @NotNull
    public static ExprNot exprNot(@NotNull Expr value) {
        return new ExprNot(value);
    }

    @NotNull
    public static ExprNullIf exprNullIf(@NotNull Expr v1, @NotNull Expr v2) {
        return new ExprNullIf(v1, v2);
    }

    @NotNull
    public static ExprOperator exprOperator(@NotNull String symbol, @Nullable Expr lhs, @NotNull Expr rhs) {
        return new ExprOperator(symbol, lhs, rhs);
    }

    @NotNull
    public static ExprOr exprOr(@NotNull Expr lhs, @NotNull Expr rhs) {
        return new ExprOr(lhs, rhs);
    }

    @NotNull
    public static ExprOverlay exprOverlay(@NotNull Expr value, @NotNull Expr placing, @NotNull Expr from, @Nullable Expr forLength) {
        return new ExprOverlay(value, placing, from, forLength);
    }

    @NotNull
    public static ExprParameter exprParameter(int index) {
        return new ExprParameter(index);
    }

    @NotNull
    public static ExprPath exprPath(@NotNull Expr root, @Nullable PathStep next) {
        return new ExprPath(root, next);
    }

    @NotNull
    public static ExprPosition exprPosition(@NotNull Expr lhs, @NotNull Expr rhs) {
        return new ExprPosition(lhs, rhs);
    }

    @NotNull
    public static ExprQuerySet exprQuerySet(@NotNull QueryBody body, @Nullable OrderBy orderBy, @Nullable Expr limit, @Nullable Expr offset) {
        return new ExprQuerySet(body, orderBy, limit, offset);
    }

    @NotNull
    public static ExprSessionAttribute exprSessionAttribute(@NotNull SessionAttribute sessionAttribute) {
        return new ExprSessionAttribute(sessionAttribute);
    }

    @NotNull
    public static ExprStruct exprStruct(@NotNull List<ExprStruct.Field> fields) {
        return new ExprStruct(fields);
    }

    @NotNull
    public static ExprStruct.Field exprStructField(@NotNull Expr name, @NotNull Expr value) {
        return new ExprStruct.Field(name, value);
    }

    @NotNull
    public static ExprSubstring exprSubstring(@NotNull Expr value, @Nullable Expr start, @Nullable Expr length) {
        return new ExprSubstring(value, start, length);
    }

    @NotNull
    public static ExprTrim exprTrim(@NotNull Expr Value, @Nullable Expr chars, @Nullable TrimSpec trimSpec) {
        return new ExprTrim(Value, chars, trimSpec);
    }

    @NotNull
    public static ExprValues exprValues(@NotNull List<ExprValues.Row> rows) {
        return new ExprValues(rows);
    }

    @NotNull
    public static ExprValues.Row exprValuesRow(@NotNull List<Expr> values) {
        return new ExprValues.Row(values);
    }

    @NotNull
    public static ExprVariant exprVariant(@NotNull String value, @NotNull String encoding) {
        return new ExprVariant(value, encoding);
    }

    @NotNull
    public static ExprVarRef exprVarRef(@NotNull IdentifierChain identifierChain, @NotNull Scope scope) {
        return new ExprVarRef(identifierChain, scope);
    }

    @NotNull
    public static ExprWindow exprWindow(@NotNull WindowFunction windowFunction, @NotNull Expr exression, @Nullable Expr offset, @Nullable Expr defaultValue, @NotNull ExprWindow.Over over) {
        return new ExprWindow(windowFunction, exression, offset, defaultValue, over);
    }

    @NotNull
    public static ExprWindow.Over exprWindowOver(@Nullable List<Expr> partitions, @Nullable List<Sort> sorts) {
        return new ExprWindow.Over(partitions, sorts);
    }

    @NotNull
    public static PathStep.Field exprPathStepField(@NotNull Identifier value, @Nullable PathStep next) {
        return new PathStep.Field(value, next);
    }

    @NotNull
    public static PathStep.Element exprPathStepElement(@NotNull Expr element, @Nullable PathStep next) {
        return new PathStep.Element(element, next);
    }

    @NotNull
    public static PathStep.AllElements exprPathStepAllElements(@Nullable PathStep next) {
        return new PathStep.AllElements(next);
    }

    @NotNull
    public static PathStep.AllFields exprPathStepAllFields(@Nullable PathStep next) {
        return new PathStep.AllFields(next);
    }

    // Graph
    @NotNull
    public static GraphLabel.Name graphLabelName(@NotNull String name) {
        return new GraphLabel.Name(name);
    }

    @NotNull
    public static GraphLabel.Wildcard graphLabelWildcard() {
        return new GraphLabel.Wildcard();
    }

    @NotNull
    public static GraphLabel.Negation graphLabelNegation(@NotNull GraphLabel arg) {
        return new GraphLabel.Negation(arg);
    }

    @NotNull
    public static GraphLabel.Conj graphLabelConj(@NotNull GraphLabel lhs, @NotNull GraphLabel rhs) {
        return new GraphLabel.Conj(lhs, rhs);
    }

    @NotNull
    public static GraphLabel.Disj graphLabelDisj(@NotNull GraphLabel lhs, @NotNull GraphLabel rhs) {
        return new GraphLabel.Disj(lhs, rhs);
    }

    @NotNull
    public static GraphMatch graphMatch(@NotNull List<GraphPattern> patterns, @Nullable GraphSelector selector) {
        return new GraphMatch(patterns, selector);
    }

    @NotNull
    public static GraphPart.Node graphMatchNode(@Nullable Expr prefilter, @Nullable String variable, @Nullable GraphLabel label) {
        return new GraphPart.Node(prefilter, variable, label);
    }

    @NotNull
    public static GraphPart.Edge graphMatchEdge(@NotNull GraphDirection direction, @Nullable GraphQuantifier quantifier, @Nullable Expr prefilter, @Nullable String variable, @Nullable GraphLabel label) {
        return new GraphPart.Edge(direction, quantifier, prefilter, variable, label);
    }

    @NotNull
    public static GraphPart.Pattern graphMatchPattern(@NotNull GraphPattern pattern) {
        return new GraphPart.Pattern(pattern);
    }

    @NotNull
    public static GraphPattern graphPattern(@Nullable GraphRestrictor restrictor, @Nullable Expr prefilter, @Nullable String variable, @Nullable GraphQuantifier quantifier, @NotNull List<GraphPart> parts) {
        return new GraphPattern(restrictor, prefilter, variable, quantifier, parts);
    }

    @NotNull
    public static GraphQuantifier graphQuantifier(long lower, @Nullable Long upper) {
        return new GraphQuantifier(lower, upper);
    }

    @NotNull
    public static GraphSelector.AnyShortest graphSelectorAnyShortest() {
        return new GraphSelector.AnyShortest();
    }

    @NotNull
    public static GraphSelector.AllShortest graphSelectorAllShortest() {
        return new GraphSelector.AllShortest();
    }

    @NotNull
    public static GraphSelector.Any graphSelectorAny() {
        return new GraphSelector.Any();
    }

    @NotNull
    public static GraphSelector.AnyK graphSelectorAnyK(long k) {
        return new GraphSelector.AnyK(k);
    }

    @NotNull
    public static GraphSelector.ShortestK graphSelectorShortestK(long k) {
        return new GraphSelector.ShortestK(k);
    }

    @NotNull
    public static GraphSelector.ShortestKGroup graphSelectorShortestKGroup(long k) {
        return new GraphSelector.ShortestKGroup(k);
    }

    // Other
    @NotNull
    public static Exclude exclude(@NotNull List<ExcludePath> excludePaths) {
        return new Exclude(excludePaths);
    }

    @NotNull
    public static ExcludePath excludePath(@NotNull ExprVarRef varRef, @NotNull List<ExcludeStep> excludeSteps) {
        return new ExcludePath(varRef, excludeSteps);
    }

    @NotNull
    public static ExcludeStep.StructField excludeStepStructField(@NotNull Identifier symbol) {
        return new ExcludeStep.StructField(symbol);
    }

    @NotNull
    public static ExcludeStep.CollIndex excludeStepCollIndex(int index) {
        return new ExcludeStep.CollIndex(index);
    }

    @NotNull
    public static ExcludeStep.StructWildcard excludeStepStructWildcard() {
        return new ExcludeStep.StructWildcard();
    }

    @NotNull
    public static ExcludeStep.CollWildcard excludeStepCollWildcard() {
        return new ExcludeStep.CollWildcard();
    }

    // This representation will be changed in https://github.com/partiql/partiql-lang-kotlin/issues/1589
    @NotNull
    public static Explain explain(@NotNull Map<String, PartiQLValue> options, @NotNull Statement statement) {
        return new Explain(options, statement);
    }

    @NotNull
    public static From from(@NotNull List<FromTableRef> tableRefs) {
        return new From(tableRefs);
    }

    @NotNull
    public static FromExpr fromExpr(@NotNull Expr expr, @NotNull FromType fromType, @Nullable Identifier asAlias, @Nullable Identifier atAlias) {
        return new FromExpr(expr, fromType, asAlias, atAlias);
    }

    @NotNull
    public static FromJoin fromJoin(@NotNull From lhs, @NotNull From rhs, @Nullable JoinType joinType, @Nullable Expr condition) {
        return new FromJoin(lhs, rhs, joinType, condition);
    }

    @NotNull
    public static GroupBy groupBy(@NotNull GroupByStrategy strategy, @NotNull List<GroupBy.Key> keys, @Nullable Identifier asAlias) {
        return new GroupBy(strategy, keys, asAlias);
    }

    @NotNull
    public static GroupBy.Key groupByKey(@NotNull Expr expr, @Nullable Identifier asAlias) {
        return new GroupBy.Key(expr, asAlias);
    }

    @NotNull
    public static Identifier identifier(@NotNull String symbol, boolean isDelimited) {
        return new Identifier(symbol, isDelimited);
    }

    @NotNull
    public static IdentifierChain identifierChain(@NotNull Identifier root, @Nullable IdentifierChain next) {
        return new IdentifierChain(root, next);
    }

    @NotNull
    public static Let let(@NotNull List<Let.Binding> bindings) {
        return new Let(bindings);
    }

    @NotNull
    public static Let.Binding letBinding(@NotNull Expr expr, @NotNull Identifier asAlias) {
        return new Let.Binding(expr, asAlias);
    }

    @NotNull
    public static OrderBy orderBy(@NotNull List<Sort> sorts) {
        return new OrderBy(sorts);
    }

    @NotNull
    public static Query query(@NotNull Expr expr) {
        return new Query(expr);
    }

    @NotNull
    public static QueryBody.SFW queryBodySFW(@NotNull Select select, @Nullable Exclude exclude, @NotNull From from, @Nullable Let let, @Nullable Expr where, @Nullable GroupBy groupBy, @Nullable Expr having) {
        return new QueryBody.SFW(select, exclude, from, let, where, groupBy, having);
    }

    @NotNull
    public static QueryBody.SetOp queryBodySetOp(@NotNull org.partiql.ast.v1.SetOp type, boolean isOuter, @NotNull Expr lhs, @NotNull Expr rhs) {
        return new QueryBody.SetOp(type, isOuter, lhs, rhs);
    }

    @NotNull
    public static SelectItem.Star selectItemStar(@NotNull Expr expr) {
        return new SelectItem.Star(expr);
    }

    @NotNull
    public static SelectItem.Expr selectItemExpr(@NotNull Expr expr, @Nullable Identifier asAlias) {
        return new SelectItem.Expr(expr, asAlias);
    }

    @NotNull
    public static SelectList selectList(@NotNull List<SelectItem> items, @Nullable SetQuantifier setq) {
        return new SelectList(items, setq);
    }

    @NotNull
    public static SelectPivot selectPivot(@NotNull Expr key, @NotNull Expr value) {
        return new SelectPivot(key, value);
    }

    @NotNull
    public static SelectStar selectStar(@Nullable SetQuantifier setq) {
        return new SelectStar(setq);
    }

    @NotNull
    public static SelectValue selectValue(@NotNull Expr constructor, @Nullable SetQuantifier setq) {
        return new SelectValue(constructor, setq);
    }

    @NotNull
    public static SetOp setOp(@NotNull SetOpType setOpType, @Nullable SetQuantifier setq) {
        return new SetOp(setOpType, setq);
    }

    @NotNull
    public static Sort sort(@NotNull Expr expr, @Nullable Order order, @Nullable Nulls nulls) {
        return new Sort(expr, order, nulls);
    }
}

File: partiql-ast/src/main/java/org/partiql/ast/v1/Ast.kt
package org.partiql.ast.v1

import org.partiql.ast.v1.expr.Expr
import org.partiql.ast.v1.expr.ExprAnd
import org.partiql.ast.v1.expr.ExprArray
import org.partiql.ast.v1.expr.ExprBag
import org.partiql.ast.v1.expr.ExprBetween
import org.partiql.ast.v1.expr.ExprCall
import org.partiql.ast.v1.expr.ExprCase
import org.partiql.ast.v1.expr.ExprCast
import org.partiql.ast.v1.expr.ExprCoalesce
import org.partiql.ast.v1.expr.ExprExtract
import org.partiql.ast.v1.expr.ExprInCollection
import org.partiql.ast.v1.expr.ExprIsType
import org.partiql.ast.v1.expr.ExprLike
import org.partiql.ast.v1.expr.ExprLit
import org.partiql.ast.v1.expr.ExprMatch
import org.partiql.ast.v1.expr.ExprNot
import org.partiql.ast.v1.expr.ExprNullIf
import org.partiql.ast.v1.expr.ExprOperator
import org.partiql.ast.v1.expr.ExprOr
import org.partiql.ast.v1.expr.ExprOverlay
import org.partiql.ast.v1.expr.ExprParameter
import org.partiql.ast.v1.expr.ExprPath
import org.partiql.ast.v1.expr.ExprPosition
import org.partiql.ast.v1.expr.ExprQuerySet
import org.partiql.ast.v1.expr.ExprSessionAttribute
import org.partiql.ast.v1.expr.ExprStruct
import org.partiql.ast.v1.expr.ExprSubstring
import org.partiql.ast.v1.expr.ExprTrim
import org.partiql.ast.v1.expr.ExprValues
import org.partiql.ast.v1.expr.ExprVarRef
import org.partiql.ast.v1.expr.ExprVariant
import org.partiql.ast.v1.expr.ExprWindow
import org.partiql.ast.v1.expr.PathStep
import org.partiql.ast.v1.expr.PathStep.AllFields
import org.partiql.ast.v1.expr.Scope
import org.partiql.ast.v1.expr.SessionAttribute
import org.partiql.ast.v1.expr.TrimSpec
import org.partiql.ast.v1.expr.WindowFunction
import org.partiql.ast.v1.graph.GraphDirection
import org.partiql.ast.v1.graph.GraphLabel
import org.partiql.ast.v1.graph.GraphMatch
import org.partiql.ast.v1.graph.GraphPart
import org.partiql.ast.v1.graph.GraphPattern
import org.partiql.ast.v1.graph.GraphQuantifier
import org.partiql.ast.v1.graph.GraphRestrictor
import org.partiql.ast.v1.graph.GraphSelector
import org.partiql.value.PartiQLValue
import org.partiql.value.PartiQLValueExperimental

// TODO docs for all factory methods
public object Ast {
    // Expr
    @JvmStatic
    public fun exprAnd(lhs: Expr, rhs: Expr): ExprAnd {
        return ExprAnd(lhs, rhs)
    }

    @JvmStatic
    public fun exprArray(values: List<Expr>): ExprArray {
        return ExprArray(values)
    }

    @JvmStatic
    public fun exprBag(values: List<Expr>): ExprBag {
        return ExprBag(values)
    }

    @JvmStatic
    public fun exprBetween(value: Expr, from: Expr, to: Expr, not: Boolean): ExprBetween {
        return ExprBetween(value, from, to, not)
    }

    @JvmStatic
    public fun exprCall(function: IdentifierChain, args: List<Expr>, setq: SetQuantifier?): ExprCall {
        return ExprCall(function, args, setq)
    }

    @JvmStatic
    public fun exprCase(expr: Expr?, branches: List<ExprCase.Branch>, defaultExpr: Expr?): ExprCase {
        return ExprCase(expr, branches, defaultExpr)
    }

    @JvmStatic
    public fun exprCaseBranch(condition: Expr, expr: Expr): ExprCase.Branch {
        return ExprCase.Branch(condition, expr)
    }

    @JvmStatic
    public fun exprCast(value: Expr, asType: DataType): ExprCast {
        return ExprCast(value, asType)
    }

    @JvmStatic
    public fun exprCoalesce(args: List<Expr>): ExprCoalesce {
        return ExprCoalesce(args)
    }

    @JvmStatic
    public fun exprExtract(field: DatetimeField, source: Expr): ExprExtract {
        return ExprExtract(field, source)
    }

    @JvmStatic
    public fun exprInCollection(lhs: Expr, rhs: Expr, not: Boolean): ExprInCollection {
        return ExprInCollection(lhs, rhs, not)
    }

    @JvmStatic
    public fun exprIsType(value: Expr, type: DataType, not: Boolean): ExprIsType {
        return ExprIsType(value, type, not)
    }

    @JvmStatic
    public fun exprLike(value: Expr, Pattern: Expr, escape: Expr?, not: Boolean): ExprLike {
        return ExprLike(value, Pattern, escape, not)
    }

    // This representation will be changed in https://github.com/partiql/partiql-lang-kotlin/issues/1589
    @OptIn(PartiQLValueExperimental::class)
    @JvmStatic
    public fun exprLit(value: PartiQLValue): ExprLit {
        return ExprLit(value)
    }

    @JvmStatic
    public fun exprMatch(expr: Expr, pattern: GraphMatch): ExprMatch {
        return ExprMatch(expr, pattern)
    }

    @JvmStatic
    public fun exprNot(value: Expr): ExprNot {
        return ExprNot(value)
    }

    @JvmStatic
    public fun exprNullIf(v1: Expr, v2: Expr): ExprNullIf {
        return ExprNullIf(v1, v2)
    }

    @JvmStatic
    public fun exprOperator(symbol: String, lhs: Expr?, rhs: Expr): ExprOperator {
        return ExprOperator(symbol, lhs, rhs)
    }

    @JvmStatic
    public fun exprOr(lhs: Expr, rhs: Expr): ExprOr {
        return ExprOr(lhs, rhs)
    }

    @JvmStatic
    public fun exprOverlay(value: Expr, placing: Expr, from: Expr, forLength: Expr?): ExprOverlay {
        return ExprOverlay(value, placing, from, forLength)
    }

    @JvmStatic
    public fun exprParameter(index: Int): ExprParameter {
        return ExprParameter(index)
    }

    @JvmStatic
    public fun exprPath(root: Expr, next: PathStep?): ExprPath {
        return ExprPath(root, next)
    }

    @JvmStatic
    public fun exprPosition(lhs: Expr, rhs: Expr): ExprPosition {
        return ExprPosition(lhs, rhs)
    }

    @JvmStatic
    public fun exprQuerySet(body: QueryBody, orderBy: OrderBy?, limit: Expr?, offset: Expr?): ExprQuerySet {
        return ExprQuerySet(body, orderBy, limit, offset)
    }

    @JvmStatic
    public fun exprSessionAttribute(sessionAttribute: SessionAttribute): ExprSessionAttribute {
        return ExprSessionAttribute(sessionAttribute)
    }

    @JvmStatic
    public fun exprStruct(fields: List<ExprStruct.Field>): ExprStruct {
        return ExprStruct(fields)
    }

    @JvmStatic
    public fun exprStructField(name: Expr, value: Expr): ExprStruct.Field {
        return ExprStruct.Field(name, value)
    }

    @JvmStatic
    public fun exprSubstring(value: Expr, start: Expr?, length: Expr?): ExprSubstring {
        return ExprSubstring(value, start, length)
    }

    @JvmStatic
    public fun exprTrim(Value: Expr, chars: Expr?, trimSpec: TrimSpec?): ExprTrim {
        return ExprTrim(Value, chars, trimSpec)
    }

    @JvmStatic
    public fun exprValues(rows: List<ExprValues.Row>): ExprValues {
        return ExprValues(rows)
    }

    @JvmStatic
    public fun exprValuesRow(values: List<Expr>): ExprValues.Row {
        return ExprValues.Row(values)
    }

    @JvmStatic
    public fun exprVariant(value: String, encoding: String): ExprVariant {
        return ExprVariant(value, encoding)
    }

    @JvmStatic
    public fun exprVarRef(identifierChain: IdentifierChain, scope: Scope): ExprVarRef {
        return ExprVarRef(identifierChain, scope)
    }

    @JvmStatic
    public fun exprWindow(
        windowFunction: WindowFunction,
        exression: Expr,
        offset: Expr?,
        defaultValue: Expr?,
        over: ExprWindow.Over
    ): ExprWindow {
        return ExprWindow(windowFunction, exression, offset, defaultValue, over)
    }

    @JvmStatic
    public fun exprWindowOver(partitions: List<Expr>?, sorts: List<Sort>?): ExprWindow.Over {
        return ExprWindow.Over(partitions, sorts)
    }

    @JvmStatic
    public fun exprPathStepField(value: Identifier, next: PathStep?): PathStep.Field {
        return PathStep.Field(value, next)
    }

    @JvmStatic
    public fun exprPathStepElement(element: Expr, next: PathStep?): PathStep.Element {
        return PathStep.Element(element, next)
    }

    @JvmStatic
    public fun exprPathStepAllElements(next: PathStep?): PathStep.AllElements {
        return PathStep.AllElements(next)
    }

    @JvmStatic
    public fun exprPathStepAllFields(next: PathStep?): AllFields {
        return AllFields(next)
    }

    // Graph
    @JvmStatic
    public fun graphLabelName(name: String): GraphLabel.Name {
        return GraphLabel.Name(name)
    }

    @JvmStatic
    public fun graphLabelWildcard(): GraphLabel.Wildcard {
        return GraphLabel.Wildcard()
    }

    @JvmStatic
    public fun graphLabelNegation(arg: GraphLabel): GraphLabel.Negation {
        return GraphLabel.Negation(arg)
    }

    @JvmStatic
    public fun graphLabelConj(lhs: GraphLabel, rhs: GraphLabel): GraphLabel.Conj {
        return GraphLabel.Conj(lhs, rhs)
    }

    @JvmStatic
    public fun graphLabelDisj(lhs: GraphLabel, rhs: GraphLabel): GraphLabel.Disj {
        return GraphLabel.Disj(lhs, rhs)
    }

    @JvmStatic
    public fun graphMatch(patterns: List<GraphPattern>, selector: GraphSelector?): GraphMatch {
        return GraphMatch(patterns, selector)
    }

    @JvmStatic
    public fun graphMatchNode(prefilter: Expr?, variable: String?, label: GraphLabel?): GraphPart.Node {
        return GraphPart.Node(prefilter, variable, label)
    }

    @JvmStatic
    public fun graphMatchEdge(
        direction: GraphDirection,
        quantifier: GraphQuantifier?,
        prefilter: Expr?,
        variable: String?,
        label: GraphLabel?
    ): GraphPart.Edge {
        return GraphPart.Edge(direction, quantifier, prefilter, variable, label)
    }

    @JvmStatic
    public fun graphMatchPattern(pattern: GraphPattern): GraphPart.Pattern {
        return GraphPart.Pattern(pattern)
    }

    @JvmStatic
    public fun graphPattern(
        restrictor: GraphRestrictor?,
        prefilter: Expr?,
        variable: String?,
        quantifier: GraphQuantifier?,
        parts: List<GraphPart>
    ): GraphPattern {
        return GraphPattern(restrictor, prefilter, variable, quantifier, parts)
    }

    @JvmStatic
    public fun graphQuantifier(lower: Long, upper: Long?): GraphQuantifier {
        return GraphQuantifier(lower, upper)
    }

    @JvmStatic
    public fun graphSelectorAnyShortest(): GraphSelector.AnyShortest {
        return GraphSelector.AnyShortest()
    }

    @JvmStatic
    public fun graphSelectorAllShortest(): GraphSelector.AllShortest {
        return GraphSelector.AllShortest()
    }

    @JvmStatic
    public fun graphSelectorAny(): GraphSelector.Any {
        return GraphSelector.Any()
    }

    @JvmStatic
    public fun graphSelectorAnyK(k: Long): GraphSelector.AnyK {
        return GraphSelector.AnyK(k)
    }

    @JvmStatic
    public fun graphSelectorShortestK(k: Long): GraphSelector.ShortestK {
        return GraphSelector.ShortestK(k)
    }

    @JvmStatic
    public fun graphSelectorShortestKGroup(k: Long): GraphSelector.ShortestKGroup {
        return GraphSelector.ShortestKGroup(k)
    }

    // Other
    @JvmStatic
    public fun exclude(excludePaths: List<ExcludePath>): Exclude {
        return Exclude(excludePaths)
    }

    @JvmStatic
    public fun excludePath(varRef: ExprVarRef, excludeSteps: List<ExcludeStep>): ExcludePath {
        return ExcludePath(varRef, excludeSteps)
    }

    @JvmStatic
    public fun excludeStepStructField(symbol: Identifier): ExcludeStep.StructField {
        return ExcludeStep.StructField(symbol)
    }

    @JvmStatic
    public fun excludeStepCollIndex(index: Int): ExcludeStep.CollIndex {
        return ExcludeStep.CollIndex(index)
    }

    @JvmStatic
    public fun excludeStepStructWildcard(): ExcludeStep.StructWildcard {
        return ExcludeStep.StructWildcard()
    }

    @JvmStatic
    public fun excludeStepCollWildcard(): ExcludeStep.CollWildcard {
        return ExcludeStep.CollWildcard()
    }

    // This representation will be changed in https://github.com/partiql/partiql-lang-kotlin/issues/1589
    @OptIn(PartiQLValueExperimental::class)
    @JvmStatic
    public fun explain(options: Map<String, PartiQLValue>, statement: Statement): Explain {
        return Explain(options, statement)
    }

    @JvmStatic
    public fun from(tableRefs: List<FromTableRef>): From {
        return From(tableRefs)
    }

    @JvmStatic
    public fun fromExpr(expr: Expr, fromType: FromType, asAlias: Identifier?, atAlias: Identifier?): FromExpr {
        return FromExpr(expr, fromType, asAlias, atAlias)
    }

    @JvmStatic
    public fun fromJoin(lhs: FromTableRef, rhs: FromTableRef, joinType: JoinType?, condition: Expr?): FromJoin {
        return FromJoin(lhs, rhs, joinType, condition)
    }

    @JvmStatic
    public fun groupBy(strategy: GroupByStrategy, keys: List<GroupBy.Key>, asAlias: Identifier?): GroupBy {
        return GroupBy(strategy, keys, asAlias)
    }

    @JvmStatic
    public fun groupByKey(expr: Expr, asAlias: Identifier?): GroupBy.Key {
        return GroupBy.Key(expr, asAlias)
    }

    @JvmStatic
    public fun identifier(symbol: String, isDelimited: Boolean): Identifier {
        return Identifier(symbol, isDelimited)
    }

    @JvmStatic
    public fun identifierChain(root: Identifier, next: IdentifierChain?): IdentifierChain {
        return IdentifierChain(root, next)
    }

    @JvmStatic
    public fun let(bindings: List<Let.Binding>): Let {
        return Let(bindings)
    }

    @JvmStatic
    public fun letBinding(expr: Expr, asAlias: Identifier): Let.Binding {
        return Let.Binding(expr, asAlias)
    }

    @JvmStatic
    public fun orderBy(sorts: List<Sort>): OrderBy {
        return OrderBy(sorts)
    }

    @JvmStatic
    public fun query(expr: Expr): Query {
        return Query(expr)
    }

    @JvmStatic
    public fun queryBodySFW(
        select: Select,
        exclude: Exclude?,
        from: From,
        let: Let?,
        where: Expr?,
        groupBy: GroupBy?,
        having: Expr?
    ): QueryBody.SFW {
        return QueryBody.SFW(select, exclude, from, let, where, groupBy, having)
    }

    @JvmStatic
    public fun queryBodySetOp(type: SetOp, isOuter: Boolean, lhs: Expr, rhs: Expr): QueryBody.SetOp {
        return QueryBody.SetOp(type, isOuter, lhs, rhs)
    }

    @JvmStatic
    public fun selectItemStar(expr: Expr): SelectItem.Star {
        return SelectItem.Star(expr)
    }

    @JvmStatic
    public fun selectItemExpr(expr: Expr, asAlias: Identifier?): SelectItem.Expr {
        return SelectItem.Expr(expr, asAlias)
    }

    @JvmStatic
    public fun selectList(items: List<SelectItem>, setq: SetQuantifier?): SelectList {
        return SelectList(items, setq)
    }

    @JvmStatic
    public fun selectPivot(key: Expr, value: Expr): SelectPivot {
        return SelectPivot(key, value)
    }

    @JvmStatic
    public fun selectStar(setq: SetQuantifier?): SelectStar {
        return SelectStar(setq)
    }

    @JvmStatic
    public fun selectValue(constructor: Expr, setq: SetQuantifier?): SelectValue {
        return SelectValue(constructor, setq)
    }

    @JvmStatic
    public fun setOp(setOpType: SetOpType, setq: SetQuantifier?): SetOp {
        return SetOp(setOpType, setq)
    }

    @JvmStatic
    public fun sort(expr: Expr, order: Order?, nulls: Nulls?): Sort {
        return Sort(expr, order, nulls)
    }
}
